shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -1554,7 +1618,84 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool, val ne
         RecordType.mk(fs)(prov)
       } else TupleType(fields.reverseIterator.mapValues(_.toUpper(noProv)))(prov)
   }
-  
+
+  def getNewVarName(prefix: String, nonValidVars: Set[Var]): String = {
+    // we check all possibe prefix_num combination, till we found one that is not in the nonValidVars
+    val ints = LazyList.from(1)
+    prefix + "_" + ints.find(index => {
+      !nonValidVars.contains(Var(prefix + "_" + index))
+    }).get
+  }
+
+  def desugarNamedArgs(term: Term, f: Term, a: Tup, argsList: List[Var])
+  (implicit ctx: Ctx, raise: Raise, vars: Map[Str, SimpleType]): SimpleType = {
+    def rec (as: List[(String -> Fld) -> Boolean], acc: Map[String, Either[Var, Term]]): Term = {
+      as match {
+        case ((v, fld), isNamed) :: tail =>
+          if (isNamed) {
+            fld.value match {
+              case _: Lit | _: Var =>
+                rec(tail, acc + (v -> R(fld.value)))
+              case _ =>
+                val newVar = Var(getNewVarName(v, a.freeVars))
+                Let(false, newVar, fld.value, rec(tail, acc + (v -> L(newVar))))
+            }
+          } else {
+            rec(tail, acc + (v -> R(fld.value)))
+          }
+        case Nil =>
+          val y: Term = Tup(argsList.map(x => 
+            acc.get(x.name) match {
+              case Some(v) =>
+                v match {
+                  case Left(v) =>
+                    (None, Fld(FldFlags.empty, v))
+                  case Right(t) => 
+                    (None, Fld(FldFlags.empty, t))
+                }
+              case None =>
+                err(s"Argument named ${x} is missing from this function call", a.toLoc)
+                (None, Fld(FldFlags.empty, Var("error")))
+            }
+          ))
+          App(f, y)
+      }
+    }
+    val hasDefined = a.fields.exists(x => x._1.isDefined)
+    val hasEmpty = a.fields.exists(x => x._1.isEmpty)
+    val areArgsMisplaced = a.fields.indexWhere(x => x._1.isDefined) < a.fields.lastIndexWhere(x => x._1.isEmpty)
+    if (hasDefined &&
+        hasEmpty && 
+        areArgsMisplaced) {
+      err("Unnamed arguments should appear first when using named arguments", a.toLoc) 
+    } else 
+      a.fields.sizeCompare(argsList) match {
+        case 0 =>
+          val as = a.fields.zipWithIndex.map{ case(x, idx) =>
+          x._1 match {
+            case Some(value) => 
+              ((value.name, x._2), true)
+            case N =>
+              ((argsList(idx).name, x._2), false)
+          }}
+          val asGroupedByVarName = as.groupBy(x => x._1._1)
+          if (asGroupedByVarName.sizeCompare(argsList) < 0) {
+            asGroupedByVarName.foreach(
+              x =>
+                x._2 match {
+                  case x1 :: y1 :: xs => err(s"Argument for parameter ${x._1} is duplicated", a.toLoc) 
+                  case _ =>     
+                }
+            )
```

Comment:
```suggestion
            asGroupedByVarName.foreach(_._2 match {
              case x1 :: y1 :: xs => err(s"Argument for parameter ${x._1} is duplicated", a.toLoc)
              case _ =>
            })
```

Reply:
Because of the use of x in error message, i think the former one is better.